src/webgpu/api/validation/vertex_access.spec.ts

```diff
@@ -47,45 +59,460 @@ TODO: make sure this isn't already covered somewhere else, review, organize, and
+    const colorAttachment = this.device.createTexture({
+      format: 'rgba8unorm',
+      size: { width: 2, height: 1, depthOrArrayLayers: 1 },
+      usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.RENDER_ATTACHMENT,
+    });
+    const colorAttachmentView = colorAttachment.createView();
+
+    const encoder = this.device.createCommandEncoder();
+    const pass = encoder.beginRenderPass({
+      colorAttachments: [
+        {
+          view: colorAttachmentView,
+          storeOp: 'store',
+          loadValue: { r: 0.0, g: 1.0, b: 0.0, a: 1.0 },
+        },
+      ],
+    });
+    return [encoder, pass];
+  }
+
+  // Generate a basic vertex buffer layout array containing given number of instance step mode and
+  // vertex step mode vertex buffer, each of them contain a given number of attributes, whose
+  // offsets form a arithmetic sequence. Within a buffer, the formats of attributes are chosen from
+  // formatList repeatedly.
+  getBasicVertexBufferLayouts({
+    arrayStride = 32,
+    attributePerBuffer = 2,
+    offsetStep = 16,
+    formatList = ['float32x4'],
+    instanceStepModeBufferCount = 1,
+    vertexStepModeBufferCount = 1,
+  }: {
+    arrayStride?: number;
+    attributePerBuffer?: number;
+    offsetStep?: number;
+    formatList?: GPUVertexFormat[];
+    instanceStepModeBufferCount?: number;
+    vertexStepModeBufferCount?: number;
+  }): GPUVertexBufferLayout[] {
+    const basicVertexBufferLayouts: GPUVertexBufferLayout[] = [];
+    let shaderLocation = 0;
+    for (
+      let buffer = 0;
+      buffer < instanceStepModeBufferCount + vertexStepModeBufferCount;
+      buffer++
+    ) {
+      const attributes: GPUVertexAttribute[] = [];
+      let offset = 0;
+      for (let attribute = 0; attribute < attributePerBuffer; attribute++) {
+        attributes.push({
+          format: formatList[attribute % formatList.length],
+          offset,
+          shaderLocation,
+        });
+        offset += offsetStep;
+        shaderLocation += 1;
+      }
+      basicVertexBufferLayouts.push({
+        arrayStride,
+        stepMode: buffer < instanceStepModeBufferCount ? 'instance' : 'vertex',
+        attributes,
+      });
+    }
+    return basicVertexBufferLayouts;
+  }
+
+  setIndexBuffers(renderEncoder: GPURenderEncoderBase, mappings: IndexBufferMapping[]) {
+    mappings.forEach(mapping =>
+      renderEncoder.setIndexBuffer(
+        mapping.buffer,
+        mapping.indexFormat,
+        mapping.offset,
+        mapping.size
+      )
+    );
+  }
+
+  setVertexBuffers(renderEncoder: GPURenderEncoderBase, mappings: VertexBufferMapping[]) {
+    mappings.forEach(mapping =>
+      renderEncoder.setVertexBuffer(mapping.slot, mapping.buffer, mapping.offset, mapping.size)
+    );
+  }
+
+  // Testing method that create render pipeline from a given buffer layout array, set index and
+  // vertex buffer before/after setting pipeline, call one given draw function if any, and check if
+  // there is any validation error.
```

Comment:
nit: This should be a `/** jsdoc comment */`, same on any other comments that describe functions.

Reply:
Done.